File: server_integration_tests/clients/ts_client/test/intercept.test.ts
import { describe, test, expect } from 'vitest';
import { newClient } from './fixtures.js';
import type { GreetingReceivedResponse } from '../src/generated/types/greeting-received-response.js';
import type { ListTeamsFieldArgs } from '../src/generated/inputs/list-teams-field-args.js';

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

describe('@glIntercept', () => {
  const client = newClient();

  test('greetIntercepted appends the runBefore marker to the name argument', async () => {
    const res = await client.queries.greetIntercepted({ name: 'Ada' });
    expect(res.greetIntercepted.message).toBe('Hello, Ada(runBefore)');
  });

  test('greetPlain is not intercepted, no marker appended', async () => {
    const res = await client.queries.greetPlain({ name: 'Ada' });
    expect(res.greetPlain.message).toBe('Hello, Ada');
  });

  test('greetingReceived (intercepted subscription) bakes the marker into every emitted greeting', async () => {
    const eventPromise = new Promise<GreetingReceivedResponse>((resolve, reject) => {
      const timer = setTimeout(() => {
        unsubscribe();
        reject(new Error('Timeout waiting for greetingReceived event'));
      }, 10000);

      const unsubscribe = client.subscriptions.greetingReceived(
        (data) => {
          clearTimeout(timer);
          unsubscribe();
          resolve(data);
        },
        (err) => {
          clearTimeout(timer);
          unsubscribe();
          reject(err);
        },
      );
    });

    // give the subscription a moment to register (and runBefore to fire) before triggering the mutation
    await sleep(300);

    await client.mutations.sendGreeting({ name: 'Ada' });

    const event = await eventPromise;
    expect(event.greetingReceived.message).toBe('Hello, Ada(runBefore)');
  });

  test('Team.members (intercepted mapping) allows role "auth"', async () => {
    const fieldArgs: ListTeamsFieldArgs = { membersRole: 'auth' };
    const res = await client.queries.listTeams({ fieldArgs });
    const members = new Set(res.listTeams[0].members.map((m) => m.name));
    expect(members).toEqual(new Set(['Ada', 'Grace']));
  });

  test('Team.members (intercepted mapping) denies any other role', async () => {
    const fieldArgs: ListTeamsFieldArgs = { membersRole: 'wrong' };
    await expect(client.queries.listTeams({ fieldArgs })).rejects.toBeTruthy();
  });
});

File: server_integration_tests/clients/dart_client/test/intercept_test.dart
import 'package:test/test.dart';
import 'package:graphlink_server_integration_tests_dart_client/generated/graphlink.dart';
import 'fixtures.dart';

void main() {
  late GraphLinkClient client;
  setUp(() => client = newClient());

  group('@glIntercept', () {
    test('greetIntercepted appends the runBefore marker to the name argument', () async {
      final res = await client.queries.greetIntercepted(name: 'Ada');
      expect(res.greetIntercepted.message, equals('Hello, Ada(runBefore)'));
    });

    test('greetPlain is not intercepted, no marker appended', () async {
      final res = await client.queries.greetPlain(name: 'Ada');
      expect(res.greetPlain.message, equals('Hello, Ada'));
    });

    test('greetingReceived (intercepted subscription) bakes the marker into every emitted greeting', () async {
      final eventFuture = client.subscriptions.greetingReceived().first.timeout(const Duration(seconds: 10));

      // give the subscription a moment to register (and runBefore to fire) before triggering the mutation
      await Future.delayed(const Duration(milliseconds: 300));

      await client.mutations.sendGreeting(name: 'Ada');

      final event = await eventFuture;
      expect(event.greetingReceived.message, equals('Hello, Ada(runBefore)'));
    });

    test('Team.members (intercepted batch mapping) allows role "auth"', () async {
      final res = await client.queries.listTeams(fieldArgs: const ListTeamsFieldArgs(membersRole: 'auth'));
      final members = res.listTeams.first.members.map((m) => m.name).toSet();
      expect(members, equals({'Ada', 'Grace'}));
    });

    test('Team.members (intercepted batch mapping) denies any other role', () async {
      final future = client.queries.listTeams(fieldArgs: const ListTeamsFieldArgs(membersRole: 'wrong'));
      await expectLater(future, throwsA(isA<List<GraphLinkError>>()));
    });
  });
}
